Platform 3901 - changing how sitemaps paginate

abador opened the pull request:
Unfortunately there wasn't a way to get rid of ORDER BY page_id (it seems they aren't in the correct order). The query that takes out all the data to generate sub sitemaps was a little bit tricky with the current query builder(the to string method returns the query for the prepare statement, without parameters), yet I decided to use the existing getQuery method and build it on top of the builder. The performance on subpages is almost linear, the sitemap index generates in about 4s, and most of this time is waiting for the db.

abador committed:
Optimize sitemap generation, PLATFORM-3901

abador committed:
Add sorting, PLATFORM-3901

abador committed:
Fix subqueries, PLATFORM-3901

abador committed:
Fix quotes, PLATFORM-3901

abador committed:
Fix quotes, PLATFORM-3901

themech commented:
Is this backwards compatible? I mean what if the main sitemap is served from the CDN (where the `$end` is not a part of the sitemap name). And the the request for the sub-sitemap gets to our backend (as for any reason it is not in the CDN). What would be the result?

abador commented:
> Is this backwards compatible? I mean what if the main sitemap is served from the CDN (where the `$end` is not a part of the sitemap name). And the the request for the sub-sitemap gets to our backend (as for any reason it is not in the CDN). What would be the result?

As far as I can tell right now the result should be a 404 http code. I can add some logic to redirect the request based on the page param. It would involve some caching for the results used by the sitemap index generator or else the redirect will take a long time.

abador committed:
Add legacy route and sitemap by page, PLATFORM-3901

abador commented:
> Is this backwards compatible? I mean what if the main sitemap is served from the CDN (where the `$end` is not a part of the sitemap name). And the the request for the sub-sitemap gets to our backend (as for any reason it is not in the CDN). What would be the result?

I've added the legacy route and sitemap generation via page param. RIght now the old routing should work. I've added some todos where the parts can be deleted after some time(when cache is refreshed)

themech commented on extensions/wikia/SitemapXml/SitemapXmlController.class.php:
```diff
@@ -114,21 +114,33 @@ private function getLastMod( $touched ) {
 	}
 
-	private function generateSitemapPage( $namespace, $page ) {
+	private function generateSitemapPage($parsedPath) {
```
Here are our code style guidelines for PHP: https://github.com/Wikia/guidelines/blob/master/PHP/CodingConventions.md. I think you need to apply those.

themech replied:
```suggestion
	private function generateSitemapPage( $parsedPath ) {
```

themech commented on extensions/wikia/SitemapXml/SitemapXmlController.class.php:
```diff
@@ -64,7 +64,7 @@ public function index() {
 			$out = $this->generateSitemapIndex();
 		} elseif ( is_int( $parsedPath->namespace ) ) {
-			$out = $this->generateSitemapPage( $parsedPath->namespace, $parsedPath->page );
+			$out = $this->generateSitemapPage( $parsedPath);
```
```suggestion
			$out = $this->generateSitemapPage( $parsedPath );
```

abador committed:
Reformat code, PLATFORM-3901

abador committed:
One more element in sitemap, PLATFORM-3901

abador committed:
Limit to int, PLATFORM-3901

abador committed:
Fox rowum mod, PLATFORM-3901

abador committed:
Fix sitemap page limits, PLATFORM-3901

abador committed:
Fix sitemap page limits, PLATFORM-3901

abador committed:
Fix build

abador committed:
Fix number of elements, PLATFORM-3901

abador committed:
Fix number of elements, PLATFORM-3901

themech commented on extensions/wikia/SitemapXml/SitemapXmlModel.class.php:
```diff
@@ -25,6 +24,36 @@ public function getPageNumber( $namespace, $limit ) {
+							->build() ) . ')' )
+				->AS_( 'db' )
+				->WHERE( 'rownum MOD ' . $limit )
+				->EQUAL_TO( '0' )
```
respect :)

abador committed:
Fix number of elements, PLATFORM-3901
